msproteomicstoolslib/format/SWATHScoringReader.py

```diff
@@ -200,6 +212,149 @@ def parse_files(self, read_exp_RT=True, verbosity=10, useCython=False):
+        self.infiles = infiles
+        self.run_id_name = "run_id"
+        self.readmethod = readmethod
+        self.aligned_run_id_name = "align_runid"
+        self.readfilter = readfilter
+        self.errorHandling = errorHandling
+        self.sequence_col = "Sequence"
+        self.read_cluster_id = read_cluster_id
+        if readmethod == "cminimal":
+            try:
+                from msproteomicstoolslib.cython.Precursor import CyPrecursor
+                from msproteomicstoolslib.cython._optimized import CyPrecursorWrapperOnly
+                self.Precursor = CyPrecursor
+                self.Precursor = CyPrecursorWrapperOnly
+            except ImportError as e:
+                print ("Requested method 'cminimal' but Cython extensions seem to be missing. Please compile and add them or use readmethod 'minimal'")
+                raise ValueError("Need Cython extensions for 'cminimal' readmethod.")
+
+        elif readmethod == "minimal":
+            self.Precursor = Precursor
+        elif readmethod == "gui":
+            self.Precursor = GeneralPrecursor
+            self.PeakGroup = GuiPeakGroup
+            self.sequence_col = "FullPeptideName"
+        else:
+            # complete
+            self.Precursor = GeneralPrecursor
+            self.PeakGroup = GeneralPeakGroup
+
+        if enable_isotopic_grouping:
+            self.peptide_group_label_name = "peptide_group_label"
+        else:
+            self.peptide_group_label_name = "transition_group_id"
+
+    def parse_file(self, filename, runs, useCython):
+        """
+        Parse a whole OSW file (which may contain data from multiple runs)
+        """
+
+        import sqlite3
+        conn = sqlite3.connect(filename)
+        c = conn.cursor()
+
+        # Retrieve and then iterate over all available runs
+        query = """SELECT ID, FILENAME FROM RUN"""
+        res = [row for row in c.execute(query)]
+        nrows = 0
+        for row in res:
+            runid = row[0]
+            current_run = Run([], {}, runid, filename, row[1], row[1], useCython=useCython)
+            runs.append(current_run)
+            nrows += self._parse_file(filename, current_run, runid, conn)
+        return nrows
+
+    def _parse_file(self, filename, run, run_id, conn):
+
+        # Make sure the correct indices exist
+        c = conn.cursor()
+        c.executescript('''
+            CREATE INDEX IF NOT EXISTS idx_precursor_precursor_id ON PRECURSOR (ID);
+            CREATE INDEX IF NOT EXISTS idx_feature_precursor_id ON FEATURE (PRECURSOR_ID);
+            CREATE INDEX IF NOT EXISTS idx_feature_feature_id ON FEATURE (ID);
+            CREATE INDEX IF NOT EXISTS idx_feature_run_id ON FEATURE (RUN_ID);
+            CREATE INDEX IF NOT EXISTS idx_feature_ms2_feature_id ON FEATURE_MS2 (FEATURE_ID);
+            CREATE INDEX IF NOT EXISTS idx_score_ms2_feature_id ON SCORE_MS2 (FEATURE_ID);
```

Comment:
@grosenberger these are the indices that we create and they should match the ones you have in pyprophet. Also, I run it multiple times in cold and hot mode, so its not the index creation that makes the difference (I think)

Reply:
Ok, maybe collecting all data from the tables just requires so much additional time. As long as it is not prohibitive even with larger data sets, I think that should not be an issue. But we should definitely check this.

Reply:
I think the problem is that it scales poorly for 1000s of datasets, e.g. if it takes several seconds to just read it into memory, it will be prohibitive serially. The issue is not the factor of 2x but the fact that there are too many peaks. 

Maybe its better to `pyprophet export` in parallel and then use the result in the alignment in serial code (smaller files, faster TSV). As you can see above, filtering the data first and then reading it in makes it ca 30x faster down to ca 1-2 seconds for 68551 peak groups (23k peak groups per file).

- 1000 seconds = 16 minutes [ 1s / file]
- 30 000 seconds = 8 hours [30 s / file]

so it clearly matters...